File: 2015/day-16/day-16-aunt-sue.js
const fs = require("fs");
const data = fs.readFileSync(__dirname + "/day-16-data.txt", "utf-8").split("\n");

const sueToMatch = {
  children: 3,
  cats: 7,
  samoyeds: 2,
  pomeranians: 3,
  akitas: 0,
  vizslas: 0,
  goldfish: 5,
  trees: 3,
  cars: 2,
  perfumes: 1,
};

function parseLine(line) {
  const itemsRegEx = /\w+\: \d+/g;
  const match = line.match(itemsRegEx);
  return {
    sueNum: +line.match(/Sue (\d+).*/)[1],
    items: match
      .map((ea) => {
        return ea.match(/(?<itemName>\w+)\: (?<itemNum>\d+)/).groups;
      })
      .reduce((out, curr) => {
        return {
          ...out,
          [curr.itemName]: +curr.itemNum,
        };
      }, {}),
  };
}

const sueData = data.map((line) => parseLine(line));

function testSue(sue) {
  let isRightSue = true;
  for (let item in sue.items) {
    if (sue.items[item] !== sueToMatch[item]) {
      isRightSue = false;
      break;
    }
  }
  return isRightSue;
}

function partTwoTestSue(sue, isPartTwo) {
  let isRightSue = true;
  for (let item in sue.items) {
    if (item === "cats" || item === "trees") {
      if (sue.items[item] <= sueToMatch[item]) {
        isRightSue = false;
        break;
      }
    } else if (item === "pomeranians" || item === "goldfish") {
      if (sue.items[item] >= sueToMatch[item]) {
        isRightSue = false;
        break;
      }
    } else if (sue.items[item] !== sueToMatch[item]) {
      isRightSue = false;
      break;
    }
  }
  return isRightSue;
}

const partOneTestedSueData = sueData.find((sue) => testSue(sue));
const partTwoTestedSueData = sueData.find((sue) => partTwoTestSue(sue));
console.log("Part 1: ", partOneTestedSueData.sueNum);
console.log("Part 2: ", partTwoTestedSueData.sueNum);

File: 2015/day-16/day-16-aunt-sue.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	f, _ := os.Open("./day-16-data.txt")
	s := bufio.NewScanner(f)

	theSue := map[string]int{
		"children":    3,
		"cats":        7,
		"samoyeds":    2,
		"pomeranians": 3,
		"akitas":      0,
		"vizslas":     0,
		"goldfish":    5,
		"trees":       3,
		"cars":        2,
		"perfumes":    1,
	}
	for s.Scan() {
		sue, sueNumStr := makeSueMap(s.Text())

		isTheSue := true
		for k, v := range sue {
			if theSue[k] != v {
				isTheSue = false
			}
		}
		if isTheSue {
			fmt.Println("Part one:", sueNumStr)
		}
		isTheSue = true
		for k, v := range sue {
			switch k {
			case "cats", "trees":
				if theSue[k] >= v {
					isTheSue = false
				}
			case "pomeranians", "goldfish":
				if theSue[k] <= v {
					isTheSue = false
				}
			default:
				if theSue[k] != v {
					isTheSue = false
				}
			}
		}
		if isTheSue {
			fmt.Println("Part two:", sueNumStr)
		}
	}
}

func makeSueMap(s string) (map[string]int, string) {
	sue := make(map[string]int)
	before, after, _ := strings.Cut(s, ": ")
	items := strings.Split(after, ", ")
	for _, i := range items {
		kv := strings.Split(i, ": ")
		key, valueString := kv[0], kv[1]
		value, _ := strconv.Atoi(valueString)
		sue[key] = value
	}
	return sue, before
}
